src/grt/src/GlobalRouter.cpp

```diff
@@ -1858,34 +1858,43 @@ void GlobalRouter::applyObstructionAdjustment(const odb::Rect& obstruction,
   }
 }
 
-// For macro pins in the east and north edges of the macros, the access for
-// them can be blocked by the macro obstructions. This function adds the
-// resources necessary to route these pins.
-// Only pins in the east and north edges are affected because FastRoute
-// routes from left to right, and bottom to top.
+// For macro pins in the east and north edges of the macros, and pins on west
+// and south edges that have their APs in GCells completely blocked by macro
+// obstructions, the access for them can be blocked by the macro obstructions.
+// This function adds the resources necessary to route these pins. Only pins in
+// the east and north edges are affected because FastRoute routes from left to
+// right, and bottom to top.
 void GlobalRouter::addResourcesForPinAccess()
 {
   odb::dbTech* tech = db_->getTech();
   for (const auto& [db_net, net] : db_net_map_) {
     for (const Pin& pin : net->getPins()) {
-      if (pin.isConnectedToPadOrMacro()
-          && (pin.getEdge() == PinEdge::east
-              || pin.getEdge() == PinEdge::north)) {
+      if (pin.isConnectedToPadOrMacro() && (pin.getEdge() != PinEdge::none)) {
         const odb::Point& pos = pin.getOnGridPosition();
         int pin_x = (int) ((pos.x() - grid_->getXMin()) / grid_->getTileSize());
         int pin_y = (int) ((pos.y() - grid_->getYMin()) / grid_->getTileSize());
         const int layer = pin.getConnectionLayer();
         odb::dbTechLayer* tech_layer = tech->findRoutingLayer(layer);
         if (tech_layer->getDirection() == odb::dbTechLayerDir::VERTICAL) {
+          const bool east_north_pin = pin.getEdge() == PinEdge::east
+                                      || pin.getEdge() == PinEdge::north;
```

Comment:
Can it happen that on a vertical layer you would be dealing with an east pin?  Likewise H and N.